odl/contrib/solvers/spdhg/misc.py

```diff
@@ -0,0 +1,857 @@
+        p: array, optional
+            initial dual variable, default=zeros
+    grad : gradient operator, optional
+        Gradient operator to be used within the total variation functional.
+        default=see TV
+    """
+
+    def __init__(self, domain, alpha=1, prox_options={}, grad=None,
+                 strong_convexity=0):
+        """
+        """
+
+        self.strong_convexity = strong_convexity
+
+        if 'name' not in prox_options:
+            prox_options['name'] = 'FGP'
+        if 'warmstart' not in prox_options:
+            prox_options['warmstart'] = True
+        if 'niter' not in prox_options:
+            prox_options['niter'] = 5
+        if 'p' not in prox_options:
+            prox_options['p'] = None
+        if 'tol' not in prox_options:
+            prox_options['tol'] = None
+
+        self.prox_options = prox_options
+
+        self.alpha = alpha
+        self.TV = TV(domain, grad=grad)
+        self.grad = self.TV.right
+        self.NN = odl.solvers.IndicatorBox(domain, 0, np.inf)
+        self.L2 = 0.5 * odl.solvers.L2NormSquared(domain)
+
+        super().__init__(space=domain, linear=False, grad_lipschitz=0)
+
+    def __call__(self, x):
+        """ Characteristic function of the non-negative orthant
+
+        Parameters
+        ----------
+        x : np.array
+            vector / image
+
+        Returns
+        -------
+        extended float (with infinity)
+            Is the input in the non-negative orthant?
+
+        Examples
+        --------
+        Check that the total variation of a constant is zero
+
+        >>> import odl.contrib.spdhg as spdhg, odl
+        >>> space = odl.uniform_discr([0, 0], [3, 3], [3, 3])
+        >>> TVNN = spdhg.TV_NonNegative(space, alpha=2)
+        >>> x = space.one()
+        >>> TVNN(x) < 1e-10
+
+        Check that negative functions are mapped to infty
+
+        >>> import odl.contrib.spdhg as spdhg, odl, numpy as np
+        >>> space = odl.uniform_discr([0, 0], [3, 3], [3, 3])
+        >>> TVNN = spdhg.TV_NonNegative(space, alpha=2)
+        >>> x = -space.one()
+        >>> np.isinf(TVNN(x))
+        """
+
+        nn = self.NN(x)
+
+        if nn is np.inf:
+            return nn
+        else:
+            out = self.alpha * self.TV(x) + nn
+            if self.strong_convexity > 0:
+                out += self.strong_convexity * self.L2(x)
+            return out
+
+    def proximal(self, sigma):
+        """ Prox operator of TV. It allows the proximal step length to be a vector
+        of positive elements.
+
+        Parameters
+        ----------
+        x : np.array
+            vector / image
+
+        Returns
+        -------
+        extended float (with infinity)
+            Is the input in the non-negative orthant?
+
+        Examples
+        --------
+        Check that the proximal operator is the identity for sigma=0
+
+        >>> import odl.contrib.spdhg as spdhg, odl, numpy as np
```

Comment:
This shouldn't be necessary (and isn't up to the `spdhg` thing). How we usually do it:

- delete this line
- add the names you want at top level, i.e. in `odl.contrib.spdhg`
- In the `run_doctests` call, supply extra global variables as `extraglobs`, in this case
  ```python
  if __name__ == '__main__':
      import odl
      import numpy as np
      from odl.contrib import spdhg
      run_doctests(extraglobs={'np': np, 'odl': odl, 'spdhg': spdhg})
  ```
  This puts the above names in the namespace of each doctest in this module.
- If you're testing a function or class that shouldn't go into the top-level namespace, just use the name directly, e.g. `TV_Nonnegative` instead of `spdhg.TV_Nonnegative`. (But I'm sure this one should go on top.)
  I know this hurts copyability to an IPython shell as @adler-j seems to have noted somewhere. But for internal helpers I don't care so much.

Reply:
The run_doctests call would go at the very end of this file?

I usually prefer if I can copy the example into my ipython shell and see the result. With the changes you proposed this would not be possible any more. Is there a way to combine the two?

Reply:
> The run_doctests call would go at the very end of this file?

Yes.



> I usually prefer if I can copy the example into my ipython shell and see the result. With the changes you proposed this would not be possible any more. Is there a way to combine the two?

You can, up to the imports. The only fully automatic way I can think of is to add all the modules to the startup code that's run when you open a new IPython shell (in Spyder). Check out Preferences -> IPython console -> Startup, you can add code to run there (multiple lines with semicolon between them).